src/gas/determineGasFeeSuggestions.ts

```diff
@@ -0,0 +1,110 @@
+import {
+  GAS_ESTIMATE_TYPES,
+  EstimatedGasFeeTimeBounds,
+  EthGasPriceEstimate,
+  GasFeeEstimates,
+  GasFeeState as GasFeeSuggestions,
+  LegacyGasPriceEstimate,
+} from './GasFeeController';
+
+/**
+ * Obtains a set of max base and priority fee estimates along with time estimates so that we
+ * can present them to users when they are sending transactions or making swaps.
+ *
+ * @param args - The arguments.
+ * @param args.isEIP1559Compatible - Governs whether or not we can use an EIP-1559-only method to
+ * produce estimates.
+ * @param args.isLegacyGasAPICompatible - Governs whether or not we can use a non-EIP-1559 method to
+ * produce estimates (for instance, testnets do not support estimates altogether).
+ * @param args.fetchGasEstimates - A function that fetches gas estimates using an EIP-1559-specific
+ * API.
+ * @param args.fetchGasEstimatesUrl - The URL for the API we can use to obtain EIP-1559-specific
+ * estimates.
+ * @param args.fetchLegacyGasPriceEstimates - A function that fetches gas estimates using an
+ * non-EIP-1559-specific API.
+ * @param args.fetchLegacyGasPriceEstimatesUrl - The URL for the API we can use to obtain
+ * non-EIP-1559-specific estimates.
+ * @param args.fetchEthGasPriceEstimate - A function that fetches gas estimates using
+ * `eth_gasPrice`.
+ * @param args.calculateTimeEstimate - A function that determine time estimate bounds.
+ * @param args.clientId - An identifier that an API can use to know who is asking for estimates.
+ * @param args.ethQuery - An EthQuery instance we can use to talk to Ethereum directly.
+ * @returns The gas fee suggestions.
+ */
+export default async function determineGasFeeSuggestions({
```

Comment:
Sure. So I was playing with `jest.mock` and I realized that it has a fundamental limitation when it comes to named exports. Say you have this file:

``` javascript
export async function makeSomeRequest() {
  return { value: 42 };
}

export async function foo() {
  const data = await makeSomeRequest();
  return data.value;
}
```

Now say you want to write a test for `foo` but you want to stub out `makeSomeRequest` (because it's a public method, so we assume it's tested too and can trust it does what it says). You might think you can say:

``` javascript
import { foo, makeSomeRequest } from './someModule';

jest.mock('./someModule');

describe('someModule', () => {
  describe('foo', () => {
    makeSomeRequest.mockResolvedValue({ value: 1 });
    const value = await foo();
    expect(value).toBe(1);
  }
})
```

But this won't work. Why? I researched this for a while and eventually came to this conclusion:

* When you want to mock a function that is exported from a module, you can't actually change that function, you have to change something that is pointing to that function.
* When you declare named exports in a module, you are essentially packaging those exports into an object and then destructuring that object when you import those exports.
* When you mock a module that has named exports, you are changing that "something" that points to the named export. This means that the `import` line will continue to work as before, but nothing actually changes within the context of the module itself (because the original function that is exported doesn't change).
* Therefore, when `foo` calls `bar` it will always call the original function — it will never call the mocked version of that function.

This is a known issue and is illustrated here: https://github.com/facebook/jest/issues/936#issuecomment-214939935

There are two ways I can think of to solve this. The first way is with a `module.exports` trick:

``` javascript
// source file

export async function makeSomeRequest() {
  return { value: 42 };
}

export async function foo() {
  const data = await module.exports.makeSomeRequest();
  return data.value;
}
```

``` javascript
// test file

import * as someModule from './someModule';

jest.mock('./someModule');

describe('someModule', () => {
  describe('foo', () => {
    someModule.makeSomeRequest.mockResolvedValue({ value: 1 });
    const value = await someModule.foo();
    expect(value).toBe(1);
  }
})
```

This works because `module.exports` in the context of `someModule.js` is the object that gets imported in the test file (under the `someModule` umbrella). And when that module changes thanks to `jest.mock`, so does `module.exports`.

If that is too obscure of a solution, another way to do this would be to change `someModule` such that instead of declaring named exports, it declares a default export:

``` javascript
// source file

async function makeSomeRequest() {
  return { value: 42 };
}

async function foo() {
  const data = await someModule.makeSomeRequest();
  return data.value;
}

const someModule = { foo, makeSomeRequest };

export default someModule;
```

``` javascript
// test file

import someModule from './someModule';

jest.mock('./someModule');

describe('someModule', () => {
  describe('foo', () => {
    someModule.makeSomeRequest.mockResolvedValue({ value: 1 });
    const value = await someModule.foo();
    expect(value).toBe(1);
  }
})
```